### PR TITLE
Implement ScatterExecutor and DistributeExecutor

### PR DESCRIPTION
**Pull request by goedderz**
Should be merged after https://github.com/arangodb/arangodb/pull/8280; this is branched off it.

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3244/

**Comment by goedderz** [comment]
All green but for a strange build failure on windows/community. Retrying, was possibly a CI failure: http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-windows-onlycatch/1236/

**Comment by goedderz** [comment]
Last windows run was also green.

File: arangod/Aql/ClusterNodes.h [hunk 1]
```diff
@@ -43,13 +43,18 @@ class IndexNode;
 class UpdateNode;
 class ReplaceNode;
 class RemoveNode;
+// TODO remove these forward declarations:
```
Comment (mchacki) [inline arangod/Aql/ClusterNodes.h]: TODO still relevant?

File: arangod/Aql/ClusterNodes.h [hunk 2]
```diff
@@ -190,7 +195,8 @@ class ScatterNode : public ExecutionNode {
 /// @brief class DistributeNode
 class DistributeNode final : public ScatterNode, public CollectionAccessingNode {
   friend class ExecutionBlock;
-  friend class DistributeBlock;
+  // TODO remove this friend, pass the necessary values in its constructor instead!
```
Comment (mchacki) [inline arangod/Aql/ClusterNodes.h]: TODO still relevant?

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3342/

**Comment by goedderz** [comment]
Jenkins is green.

**Comment by mchacki** [comment]
After merge jenkins: 
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3374/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3480/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3544/

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3548/

**Comment by mchacki** [comment]
The Mac  catch test has been fixed upstream and is unrelated

**Pull request merged**